### ISSUE TITLE
throttle and debounce are broken on master (ok on CRAN)

### ISSUE THREAD

**Issue by maxheld83** [rstudio/shiny#2725]
I think `debounce()` and `throttle()` are broken as of https://github.com/rstudio/shiny/commit/e84beffee39d540d3e891faafb9f333af32deca4.

Here's a reprex, taken from the [docs](https://shiny.rstudio.com/reference/shiny/1.4.0/debounce.html):

```
library(shiny)
library(magrittr)

ui <- fluidPage(
  plotOutput("plot", click = clickOpts("hover")),
  helpText("Quickly click on the plot above, while watching the result table below:"),
  tableOutput("result")
)

server <- function(input, output, session) {
  hover <- reactive({
    if (is.null(input$hover))
      list(x = NA, y = NA)
    else
      input$hover
  })
  hover_d <- hover %>% debounce(1000)
  hover_t <- hover %>% throttle(1000)

  output$plot <- renderPlot({
    plot(cars)
  })

  output$result <- renderTable({
    data.frame(
      mode = c("raw", "throttle", "debounce"),
      x = c(hover()$x, hover_t()$x, hover_d()$x),
      y = c(hover()$y, hover_t()$y, hover_d()$y)
    )
  })
}

shinyApp(ui, server)
}

```

With CRAN (1.4.0) (`install.packages('shiny)`), the throttled/debounced coordinates "lag behind", as expected.

With `master` (`remotes::install_github('rstudio/shiny'), the throttled/debounced coordinates are in near-perfect sync with the unthrottled coordinates. 
The remaining lag is probably just shiny at work, and in any case, does not change when `millis` is changed.

In my usage outside of the reprex, it seemed that for the *first* invalidation, the throttle/debounce is respected, but not for any later ones.
Can't reproduce that behavior in the reprex, probably because the first click is so fast.
Perhaps this helpful in debugging.

----

`devtools::session_info()`.

```
─ Session info ─────────────────────────────────────────────────────────────────────
 setting  value                       
 version  R version 3.6.1 (2019-07-05)
 os       macOS Catalina 10.15.1      
 system   x86_64, darwin15.6.0        
 ui       RStudio                     
 language en_US.UTF-8 git             
 collate  en_US.UTF-8                 
 ctype    en_US.UTF-8                 
 tz       Europe/Berlin               
 date     2019-12-10                  

─ Packages ─────────────────────────────────────────────────────────────────────────
 package     * version    date       lib source                            
 assertthat    0.2.1      2019-03-21 [1] CRAN (R 3.6.0)                    
 backports     1.1.5      2019-10-02 [1] CRAN (R 3.6.0)                    
 callr         3.3.2      2019-09-22 [1] CRAN (R 3.6.0)                    
 cli           1.1.0      2019-03-19 [1] CRAN (R 3.6.0)                    
 crayon        1.3.4      2017-09-16 [1] CRAN (R 3.6.0)                    
 desc          1.2.0      2018-05-01 [1] CRAN (R 3.6.0)                    
 devtools      2.2.1      2019-09-24 [1] CRAN (R 3.6.1)                    
 digest        0.6.23     2019-11-23 [1] CRAN (R 3.6.0)                    
 ellipsis      0.3.0      2019-09-20 [1] CRAN (R 3.6.0)                    
 evaluate      0.14       2019-05-28 [1] CRAN (R 3.6.0)                    
 fastmap       1.0.1      2019-10-08 [1] CRAN (R 3.6.0)                    
 fs            1.3.1      2019-05-06 [1] CRAN (R 3.6.0)                    
 glue          1.3.1      2019-03-12 [1] CRAN (R 3.6.0)                    
 htmltools     0.4.0      2019-10-04 [1] CRAN (R 3.6.0)                    
 httpuv        1.5.2      2019-09-11 [1] CRAN (R 3.6.1)                    
 knitr         1.26       2019-11-12 [1] CRAN (R 3.6.1)                    
 later         1.0.0      2019-10-04 [1] CRAN (R 3.6.0)                    
 magrittr    * 1.5        2014-11-22 [1] CRAN (R 3.6.0)                    
 memoise       1.1.0      2017-04-21 [1] CRAN (R 3.6.0)                    
 mime          0.7        2019-06-11 [1] CRAN (R 3.6.0)                    
 packrat       0.5.0      2018-11-14 [1] CRAN (R 3.6.0)                    
 pkgbuild      1.0.6      2019-10-09 [1] CRAN (R 3.6.0)                    
 pkgload       1.0.2      2018-10-29 [1] CRAN (R 3.6.0)                    
 prettyunits   1.0.2      2015-07-13 [1] CRAN (R 3.6.0)                    
 processx      3.4.1      2019-07-18 [1] CRAN (R 3.6.0)                    
 promises      1.1.0      2019-10-04 [1] CRAN (R 3.6.0)                    
 ps            1.3.0      2018-12-21 [1] CRAN (R 3.6.0)                    
 R6            2.4.1      2019-11-12 [1] CRAN (R 3.6.1)                    
 Rcpp          1.0.3      2019-11-08 [1] CRAN (R 3.6.0)                    
 remotes       2.1.0      2019-06-24 [1] CRAN (R 3.6.0)                    
 rlang         0.4.2      2019-11-23 [1] CRAN (R 3.6.0)                    
 rmarkdown     1.18       2019-11-27 [1] CRAN (R 3.6.0)                    
 rprojroot     1.3-2      2018-01-03 [1] CRAN (R 3.6.0)                    
 rstudioapi    0.10       2019-03-19 [1] CRAN (R 3.6.0)                    
 sessioninfo   1.1.1      2018-11-05 [1] CRAN (R 3.6.0)                    
 shiny       * 1.4.0      2019-10-10 [1] CRAN (R 3.6.0)                    
 testthat      2.3.1      2019-12-01 [1] CRAN (R 3.6.0)                    
 usethis       1.5.1.9000 2019-08-22 [1] Github (jimhester/usethis@b25c982)
 withr         2.1.2      2018-03-15 [1] CRAN (R 3.6.0)                    
 xfun          0.11       2019-11-12 [1] CRAN (R 3.6.1)                    
 xtable        1.8-4      2019-04-21 [1] CRAN (R 3.6.0)                    
 yaml          2.2.0      2018-07-25 [1] CRAN (R 3.6.0)                    

[1] /Library/Frameworks/R.framework/Versions/3.6/Resources/library
``

**Comment by wch** [comment]
Thanks for the great bug report! I used a slightly modified version of the test app, which actually uses hovering (I don't know why the doc uses clicks when the var name is `hover`):

```R
library(shiny)
library(magrittr)

ui <- fluidPage(
  plotOutput("plot", hover = "hover"),
  helpText("Quickly click on the plot above, while watching the result table below:"),
  tableOutput("result")
)

server <- function(input, output, session) {
  hover <- reactive({
    if (is.null(input$hover))
      list(x = NA, y = NA)
    else
      input$hover
  })
  hover_d <- hover %>% debounce(1000)
  hover_t <- hover %>% throttle(1000)

  output$plot <- renderPlot({
    plot(cars)
  })

  output$result <- renderTable({
    data.frame(
      mode = c("raw", "throttle", "debounce"),
      x = c(hover()$x, hover_t()$x, hover_d()$x),
      y = c(hover()$y, hover_t()$y, hover_d()$y)
    )
  })
}

shinyApp(ui, server)
```

I bisected it to d6b46f82435d0effc03ea98108e66fdedbb80ba2.

**Comment by wch** [comment]
I think the problem is that the `getTime()` function in some cases returns `Sys.time()`, and other cases returns `as.numeric(Sys.time()) * 1000`.

**Comment by wch** [comment]
@maxheld83 Can you try out the fix in #2728?

```R
remotes::install_github('rstudio/shiny#2728')
```

**Comment by maxheld83** [comment]
I can confirm that #2728 resolves the issue both in the above reprex and my own app
